addon/-private/router-ext.js

```diff
@@ -0,0 +1,202 @@
+import Ember from 'ember';
+import emberRequire from './ext-require';
+
+const hasDefaultSerialize = emberRequire('ember-routing/system/route', 'hasDefaultSerialize');
+
+const {
+  Logger: {
+    info
+  },
+  Router,
+  RSVP,
+  assert,
+  get,
+  getOwner
+} = Ember;
+
+Router.reopen({
+  assetLoader: Ember.inject.service(),
+
+  /**
+   * We skip doing default query params stuff when attempting to go to an
+   * Engine route.
+   *
+   * @override
+   */
+  _prepareQueryParams(routeName) {
+    if (this._engineInfoByRoute[routeName]) {
+      return;
+    }
+
+    return this._super(...arguments);
+  },
+
+  /**
+   * We override this to fetch assets when crossing into an Engine. For other
+   * cases we do the normal thing.
+   *
+   * @override
+   */
+  _getHandlerFunction() {
+    let seen = Object.create(null);
```

Comment:
I'm not sure if we want to use an object here. I believe this object is going to be prone to a lot of cache misses and would also be susceptible to blowing up the IC at site where we check the cache. I may be wrong so @krisselden should probably comment sense he's been working in this domain.